### PR TITLE
Remove dependency on argv[0]

### LINKED ISSUE #431: Remove the need for `argv[0]` being the app name

### PR DESCRIPTION
**Pull request by abouteiller**
Remove the dependency on ARGV[0] by getting the app name from somewhere else. 

- [x] change the examples to not copy argv0 everywhere now that it is unnecessary. 
- [x] test on OSX
- [x] test on windows WSL
- [ ] test on windows Visual Studio/native

fixes #431

### CODE REVIEW

**Review by bosilca — APPROVED** [review]
Nice solutions.

File: tests/apps/haar_tree/main.c [hunk 1]
```diff
@@ -171,20 +171,9 @@ int main(int argc, char *argv[])
             break;
         }
     }
-    pargv = malloc( (pargc+1) * sizeof(char*));
```
Comment (therault) [inline tests/apps/haar_tree/main.c:174]: Wait, we still need to find "--" on the command line to pass it to parsec_init: arguments before "--" are not understood by parsec_init and will create an error message.

File: tests/apps/haar_tree/main.c [hunk 2]
```diff
@@ -171,20 +171,9 @@ int main(int argc, char *argv[])
...
-            dashdash = i;
-            pargc = 0;
-        } else if( dashdash != -1 ) {
-            pargc++;
-        }
-    } else {
-        pargv[0] = NULL;
-    }
     parsec = parsec_init(1, &pargc, &pargv);
     free(pargv);
```
Comment (therault) [inline tests/apps/haar_tree/main.c:175]: Also, you kept free(pargv), but pargv is not malloc anymore?

**Review by therault — CHANGES_REQUESTED** [review]
I like the `process_name()` solution, but there is one example broken.

The other example fixes are correct, AFAICT

File: parsec/parsec.c [hunk 3]
```diff
@@ -705,7 +701,7 @@ parsec_context_t* parsec_init( int nb_cores, int* pargc, char** pargv[] )
             /* Specialize the profiling filename to avoid collision with other instances */
             ret = asprintf( &cmdline_info, "%s_%d", basename(parsec_app_name), (int)getpid() );
             if (ret < 0) {
-                cmdline_info = strdup(DEFAULT_APP_NAME);
+                cmdline_info = strdup(parsec_app_name);
```
Comment (bosilca) [inline parsec/parsec.c:704]: parsec_app_name is already a malloced string. Do you need to strdup it here ?

File: parsec/utils/process_name.c [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    }
+#else
+    ret = readlink("/proc/self/exe", name, len);
+    if(-1 == ret) {
+        snprintf(name, len, "parsec-app" );
+        return strdup(name);
+    }
+    name[ret] = '\0';
+#endif
+    sname = rindex(name, '/');
```
Comment (bosilca) [inline parsec/utils/process_name.c:25]: The path separator in parsec is `PARSEC_PATH_SEP`

**Review by bosilca — CHANGES_REQUESTED** [review]

File: parsec/utils/process_name.c (hunk 4 above)
Comment (abouteiller) [reply parsec/utils/process_name.c:25]: done

File: tests/apps/haar_tree/main.c (hunk 1 above)
Comment (abouteiller) [reply tests/apps/haar_tree/main.c:174]: yes look at line 168 above

**Review by bosilca — APPROVED** [review]
Here is a [Stackoverflow](https://stackoverflow.com/questions/4570174/how-to-get-the-process-name-in-c#:~:text=You%20can%20obtain%20the%20process,you%20can%20also%20use%20GetProcessImageFileName.) possible Windows solution

**Comment by abouteiller** [comment]
I tried with WSL and it doesn't require that trick. I couldn't compile with Visual Studio (can't find MPI for some reason). Do we really care about something else than WSL at this point?

**Comment by bosilca** [comment]
Let's merge it.

**Pull request merged**